pkg/storegateway/bucket_index_reader.go

```diff
@@ -32,18 +32,35 @@ import (
 
 // expandedPostingsPromise is the promise returned by bucketIndexReader.expandedPostingsPromise.
 // The second return value indicates whether the returned data comes from the cache.
-type expandedPostingsPromise func(ctx context.Context) ([]storage.SeriesRef, bool, error)
+type expandedPostingsPromise func(ctx context.Context) ([]storage.SeriesRef, []*labels.Matcher, bool, error)
+
+type postingsSelectionStrategy interface {
+	name() string
+	selectPostings([]postingGroup) (selected, omitted []postingGroup)
+}
+
+type selectAllStrategy struct{}
+
+func (selectAllStrategy) name() string {
+	return "selectAll"
+}
+
+func (selectAllStrategy) selectPostings(groups []postingGroup) (selected, omitted []postingGroup) {
+	return groups, nil
+}
 
 // bucketIndexReader is a custom index reader (not conforming index.Reader interface) that reads index that is stored in
 // object storage without having to fully download it.
 type bucketIndexReader struct {
-	block *bucketBlock
-	dec   *index.Decoder
+	block         *bucketBlock
+	postingsStrat postingsSelectionStrategy
```

Comment:
I'd go with `ps postingsSelectonStrategy` or `postingsSelectonStrategy postingsSelectonStrategy`, but that shortening is weird. IMO. Feel free to ignore this.